Addons: sorting algorithm for versions customizable on flyout

humitos opened the pull request:
Allow users to choose one of the pre-defined algorithms:

- Lexicographically
- SemVer (Read the Docs)
- CalVer
- Custom pattern

The sorting algorithm is implemented in the backend. So, the list returned under `addons.flyout.versions` will be sorted acordingly the algorithm the user chose. There is no need to do anything extra in the front-end.

The algorithm follows the next general rule: _"all the versions that don't match the pattern defined, are considered invalid and sorted lexicographically between them and added to the end of the list"_. That means that valid versions will appear always first in the list and sorted together with other _valid versions_.

There is one _key feature_ here and is that the user can define any pattern supported by `bumpver`, which is the module we use behind the scenes to perform the sorting. See https://github.com/mbarkhau/bumpver#pattern-examples

On the other hand, `SemVer (Read the Docs)` is implemented used the exact same code we were using for the old flyout implementation. This is mainly to keep backward compatibility, but its usage is not recommended since it's pretty hard to explain to users and hide non-clear/weird behavior.

-----

## Screenshots

![Screenshot_2024-01-26_14-33-51](https://github.com/readthedocs/readthedocs.org/assets/244656/5fbb017a-3bef-4c69-8d80-2651027b2bd6)


![Screenshot_2024-01-26_14-34-15](https://github.com/readthedocs/readthedocs.org/assets/244656/09e33d6d-fbc4-48e9-ba11-c7cfb1be2d54)

## ToDo

- [ ] write simple test cases for each algorithm
- [ ] write test cases for project's form




Closes https://github.com/readthedocs/addons/issues/222

ericholscher commented on readthedocs/projects/models.py:
```diff
@@ -179,6 +181,17 @@ class AddonsConfig(TimeStampedModel):
+        default=None,
+        null=True,
+        blank=True,
+        help_text="Sorting pattern supported by BumpVer",
```
This should probably be a link, if we're going to call out `BumpVer` explicitly.

humitos replied:
Yes. I think we will eventually document this by ourselves, but I think it's fine for now to provide a link to users so they know where to find out this information.

ericholscher reviewed (commented):
This is a nice pattern. I think we'll want to break this out into its own heading in the Addons settings to make it a bit nicer to configure. I know there's other conversation around that, but seems more important as we add options here.

humitos commented:
I'm not sure where my comment went, so I'm writing it again 🤷🏼 

---

> I think we'll want to break this out into its own heading in the Addons settings to make it a bit nicer to configure. I know there's other conversation around that, but seems more important as we add options here.


I suggested splitting this page into one section per addon in https://github.com/readthedocs/readthedocs.org/pull/11031, but I used a CrispyForm pattern we want to avoid. @agjohnson said he will have more time in the following days to take a look at that and communicate what's the pattern we want to follow there. Once that's defined, I'm happy to jump into the work to implement this pattern on this page 👍🏼

stsewd commented on readthedocs/projects/version_handling.py:
```diff
@@ -158,3 +161,86 @@ def determine_stable_version(version_list):
+    """
+    alphabetically_sorted_version_list = sorted(
+        version_list,
+        key=operator.attrgetter("slug"),
```
We should probably use `verbose_name` for sorting.

humitos replied:
Hrm... Good point! I would keep this change for another PR because I've been using slug in multiple places on the addons API (example: https://github.com/readthedocs/readthedocs.org/blob/96fc90bfdc318c962e14c6e92da16bd55cd176b5/readthedocs/proxito/views/hosting.py#L333-L341) and we are using `slug` to compare things in the addons JavaScript as well: https://github.com/readthedocs/readthedocs-client/blob/7a1735049289d1c8c7d706f741aa092e1190d5ec/src/flyout.js#L199-L206

So, we would need to change all of these places and it will require a bigger change here.

humitos replied:
I created an issue for this at https://github.com/readthedocs/readthedocs.org/issues/11182

stsewd reviewed (commented):
- We should use the verbose name of the version to sort, not the slug.
- Shouldn't we still list latest and stable first? Those are special versions, basically an alias for the two mayor versions.
- Talking about latest and stable, shouldn't we use these same algorithms to get the stable version?

agjohnson reviewed (commented):
This is a great feature addition, I know we have had this requested in the past.

I think this form looks okay for now, but would keep this UI mind for a later upgrade too. It is difficult and somewhat confusing trying to explain these options just using the option text description. One small upgrade might be to expose the underlying pattern for each as a read only field (similar to the project redirect UI), or even exposing a description and/or example for each option. Both of these would be JS/Knockout additions, and out of scope for a first pass though.

But for that, I wouldn't try to explain the options using the option text description either.

humitos commented:
> Shouldn't we still list latest and stable first? Those are special versions, basically an alias for the two mayor versions.

This makes sense to me. I will give it a try. This may also be configurable, like "Show `latest` and `stable` at the beginning (default `True`)" and users can disable it if they don't like that approach.

> Talking about latest and stable, shouldn't we use these same algorithms to get the stable version?

I'd say yes, but that's a _big_ refactor 😄

humitos commented on readthedocs/projects/constants.py:
```diff
@@ -431,3 +431,23 @@
+ADDONS_FLYOUT_SORTING_CUSTOM_PATTERN = "custom-pattern"
+
+ADDONS_FLYOUT_SORTING_CHOICES = (
+    (ADDONS_FLYOUT_SORTING_LEXICOGRAPHYCALLY, "Alphabetically"),
+    (ADDONS_FLYOUT_SORTING_SEMVER_READTHEDOCS_COMPATIBLE, "SemVer (Read the Docs)"),
+    (ADDONS_FLYOUT_SORTING_PYTHON_PACKAGING, "Python Packaging (PEP 440 and PEP 425)"),
+    (ADDONS_FLYOUT_SORTING_CALVER, "CalVer (YYYY.0M.0M)"),
+    (ADDONS_FLYOUT_SORTING_CUSTOM_PATTERN, "Define your own pattern"),
```
He, we _need_ a differentiation from the original SemVer itself because our algorithm is a pretty customized version of SemVer. I kept it here because is the current algorithm we are using in the old flyout and will help with the migration, but we need to differentiate it from the _real SemVer_.

> I don't think users will known what `(Read the Docs)` implies here.

We need documentation for each of these algorithms.

agjohnson replied:
Yeah, so I wouldn't worry much about describing the options in the option text then. I would do that in documentation. I also feel more complex UI would really solve usability issues here too, but I'd want to see a fair amount of usage here first.

agjohnson commented:
> Shouldn't we still list latest and stable first?

I'd agree that latest/stable at the top of the listing should be preserved though. It won't always look out of place, but depending on the branch/tag naming the ordering could be a bit janky, ie:

- dev
- foo
- **latest**
- something
- something
- stable

This could be a case for trying to find a different place in the flyout UI/UX to describe latest and stable too.

----

Just in case there is some overlap here, I ran into this same issue on the version listing dashboard view. It would be great if the version listing filter had a default sort option that matched the documentation listing order, and showed latest/stable up front.

Either reusing the code in this PR, or eventually working this into the version filter and reusing that in the API could be options. Just noting for now though.

humitos commented:
> This could be a case for trying to find a different place in the flyout UI/UX to describe latest and stable too.

This is definitely something that will be pretty clear to users. I would consider this when redesigning the flyout in the next iteration. In the meantime, I think we can implement the simple behavior proposed by Santos that keeps `latest` and `stable` at the beginning with an option to opt-out.

Should it be `stable   latest   ... ` or `latest   stable   ...`? What is the version we want the users to click on most of the times? cc @agjohnson @stsewd

stsewd commented:
I think it makes sense to list latest first and then stable, that's how they will be sorted if we use their aliases.

humitos commented:
I updated this PR to bring latest changes and made strings translatable. I also implemented `latest` and `stable` at the beginning of the list. The only missing thing here is to write some tests as I put in the description. Once that's done, we can move forward with the review and merge of it.

humitos commented:
CircleCI `checks` fails because a new requirements is needed, but can't be installed there until the PR gets merged 🙃 

```
ModuleNotFoundError: No module named 'bumpver'
```

It's fine to move forward, tho 👍🏼

humitos commented on readthedocs/projects/tests/test_version_handling.py:
```diff
@@ -0,0 +1,207 @@
+        ]
+
+        expected = [
+            "v1.0",
+            "1.1",
+            "1.1.0",
+            "2.5.3",
```
Pinging @astrojuanlu since you are one of the users wanting this feature.

astrojuanlu replied:
Yeah I think descending order makes more sense

humitos replied:
OK, the result for this case using Python Packaging sorting will be:

```
2.5.3	1.1	1.1.0	v1.0
```

Then, the final result with Read the Docs special versions and _invalid ones_ will be:

```
latest	stable	2.5.3	1.1	1.1.0	v1.0	another	invalid
```

The pattern is:

```
<latest if active> <stable if active> <versions matching the pattern descending> <invalid versions alphabetically>
```

ericholscher commented on readthedocs/projects/forms.py:
```diff
@@ -588,6 +589,9 @@ class Meta:
             "flyout_enabled",
+            "flyout_sorting",
+            "flyout_sorting_latest_stable_at_beginning",
+            "flyout_sorting_custom_pattern",
```
Are we doing any logic to make this show/hide based on the value? I don't see it in this PR.

humitos replied:
Yes. We will need to work on this in the front-end (ext-theme) to show/hide them dynamically using KO based on what the user selects on flyout sorting.

humitos commented on readthedocs/projects/version_handling.py:
```diff
@@ -158,3 +167,106 @@ def determine_stable_version(version_list):
+    )
+
+
+def sort_versions_custom_pattern(version_list, raw_pattern, latest_stable_at_beginning):
```
Hrm, this is a good point.

I took a look at this and it's not super easy to break it out into a common function, but maybe writing a class that override the logic inside try/except (which is the only difference) could be easier. I'm not sure yet, tho, but I will give it a try.

humitos replied:
I opened https://github.com/readthedocs/readthedocs.org/issues/11207 to com back to this in another PR.
